contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,555 @@
+        require(
+            transformComponentInfo[_transformComponent].underlyingComponent != address(0),
+            "TransformInfo not set yet"
+        );
+
+        // Maximum transform size is limited to MAX_UINT_96 to prevent overflows when multiplying this maximum size by
+        // the transform components exchange rate to get a maximum untransform size. This prevents using MAX_UINT_256 
+        // to represent an unlimited amount.
+        require(_transformInfo.maxTransformSize <= type(uint96).max, "max transform size must be less than MAX_UINT_96");
+        transformComponentInfo[_transformComponent] = _transformInfo;
+    }
+
+    /**
+     * ONLY OPERATOR: Begins a rebalance. Must supply the set components as well as the target unit allocation for each. If the
+     * component is a transform component, then the target units should be measured in the equivalent value of the underlying tokens
+     *
+     * @param _setComponents            array of components involved in rebalance including components being removed (target units set to 0)
+     * @param _targetUnitsUnderlying    array of target units at end of rebalance, maps to same index of _components array
+     */
+    function startIPRebalance(address[] memory _setComponents, uint256[] memory _targetUnitsUnderlying) external onlyOperator {
+        require(_setComponents.length == _targetUnitsUnderlying.length, "length mismatch");
+        stage = RebalanceStage.UNTRANSFORM;
+
+        // Save rebalanceParams
+        for (uint256 i = 0; i < _setComponents.length; i++) {
+            targetUnitsUnderlying[_setComponents[i]] = _targetUnitsUnderlying[i];
+        }
+
+        setComponentList = _setComponents;
+    }
+
+    /**
+     * ONLY ALLOWED CALLER: Untransforms components. This function must be called after starting the rebalance, but before calling startTrades.
+     * The untransformData parameter can be fetched by the rebalance bots by calling getUntransformData on the relevent TransformHelper.
+     *
+     * @param _transformComponents      array of components to untransform
+     * @param _untransformData          array of untransform data 
+     */
+    function batchUntransform(
+        address[] memory _transformComponents,
+        bytes[] memory _untransformData
+    )
+        external
+        onlyAllowedCaller(msg.sender)
+    {
+        require(_transformComponents.length == _untransformData.length, "length mismatch");
+
+        _absorbAirdrops(_transformComponents);
+
+        for (uint256 i = 0; i < _transformComponents.length; i++) {
+            _untransform(_transformComponents[i], _untransformData[i]);
+        }
+    }
+
+    /**
+     * ONLY OPERATOR: Begins the trading portion of the rebalance by initializing the GeneralIndexModule. Must be called after all required
+     * components have been untransformed.
+     */
+    function startTrades() external onlyOperator {
+        stage = RebalanceStage.TRADE;
```

Comment:
Let's set the TRADE state after successful completion of untransform functions, and verify we are in the TRADE state before executing startTrades.

Reply:
See https://github.com/SetProtocol/index-coop-smart-contracts/pull/81#discussion_r773476469